### PR TITLE
24 nagios plugin launcher

### PR DESCRIPTION
**Pull request by ghost**
Initial check launcher. Creates a process for each check using multiprocessing, while each process has subprocess to work with output. Also sync with database is performed in order to change service and check status according to output result after calling command bounded to check.

### CODE REVIEW

File: moninag/daemons/check_launcher.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
+"""This module provides methods for launching check commands as multiple processes and synchronizing them to database"""
```
Comment (randombluff) [inline moninag/daemons/check_launcher.py]: line too long (120 > 100 characters)

File: moninag/daemons/check_launcher.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+# Path to project directory where manage.py is located
+PROJECT_PATH = '../'
+sys.path.append(PROJECT_PATH)
+
+# This is so Django knows where to find stuff
+os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'moninag.settings')
+django.setup()
+
+from check.models import Check
```
Comment (randombluff) [inline moninag/daemons/check_launcher.py:25]: module level import not at top of file ;) why is it here?

File: moninag/daemons/check_launcher.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    check_keys = set(checks.in_bulk().keys())
+
+    # Create a dict of checks which has check_id, command and run_frequency
+    check_dict = {}
+    for check in checks:
+        check_dict[check.id] = {}
+
+        # if this check needs port to run the command attach target_port to the command
+        if '{port}' in check.plugin.template:
+            check_dict[check.id]['command'] = check.plugin.template.format(host=check.service.server.address,
```
Comment (randombluff) [inline moninag/daemons/check_launcher.py]: line too long (109 > 100 characters)

File: moninag/daemons/check_launcher.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+                                                                           port=check.target_port)
+        else:
+            check_dict[check.id]['command'] = check.plugin.template.format(host=check.service.server.address)
```
Comment (randombluff) [inline moninag/daemons/check_launcher.py]: line too long (109 > 100 characters)

File: moninag/daemons/check_launcher.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        new_storage = copy.deepcopy(storage)
+
+        # Fetch data from database
+        check_sync(new_storage)
+
+        # Get storage keys in order to compare storages for changes
+        old_keys = set(storage.keys())
+        new_keys = set(new_storage.keys())
+
+        # Get keys of added_checks/deleted_checks/common_checks elements in init storage and updated storage
```
Comment (randombluff) [inline moninag/daemons/check_launcher.py]: line too long (108 > 100 characters)

File: moninag/daemons/collectordaemon.py [hunk 6]
```diff
@@ -58,14 +58,17 @@ def perform_task(self):
...
+                                  '(Service #:{service_id}) (Last Run: {last_run}) '
                                   '(Plugin Name: {pluginname}) (Target Port: {targ_port}) '
-                                  '(Run Frequency: {run_freq}sec) '.format(id=check.id,
-                                                                           name=check.name,
-                                                                           service_id=check.service.id,
-                                                                           pluginname=check.plugin_name,
-                                                                           targ_port=check.target_port,
-                                                                           run_freq=check.run_freq))
+                                  '(Run Frequency: {run_freq}sec)'.format(id=check.id,
+                                                                          name=check.name,
```
Comment (maaraanas) [inline moninag/daemons/collectordaemon.py]: line too long (102 > 100 characters)

File: moninag/daemons/check_launcher.py (hunk 2 above)
Comment (ghost) [reply moninag/daemons/check_launcher.py:25]: since we cant reach it without code which is above this import

File: moninag/check/models.py [hunk 7]
```diff
@@ -66,6 +72,24 @@ def update(self, name=None, plugin=None, target_port=None, run_freq=None):
...
+        This method updates service status depending on this services checks statuses
+        """
+
+        checks = Check.objects.filter(service=self.service.id)
+        statuses = [check.status for check in checks]
+        if all(status == 'OK' for status in statuses):
+            self.service.status = 'Ok'
+
+        elif 'Fail' or 'Warning' in statuses:
+            if not 'Fail' in statuses:
```
Comment (hutsulyk) [inline moninag/check/models.py]: Test for membership should be `not in`.

File: moninag/check/models.py [hunk 8]
```diff
@@ -66,6 +72,24 @@ def update(self, name=None, plugin=None, target_port=None, run_freq=None):
...
+                self.service.status = 'Warning'
+            else:
+                self.service.status = 'Fail'
```
Comment (hutsulyk) [inline moninag/check/models.py]: Maybe it is better to use lower case for statuses and comparison?

File: moninag/daemons/check_launcher.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    if b'OK' in output_list[1]:
+        # Set check status to OK
+        this_check.status = 'OK'
+
+    elif b'WARNING' in output_list[1]:
+        # Set check status to WARNING
+        this_check.status = 'Warning'
+    else:
+        # Set check status to Fail
+        this_check.status = 'Fail'
```
Comment (hutsulyk) [inline moninag/daemons/check_launcher.py]: Update here too with Uppercase :slightly_smiling_face:

File: moninag/check/models.py [hunk 10]
```diff
@@ -66,6 +73,24 @@ def update(self, name=None, plugin=None, target_port=None, run_freq=None):
...
+        checks = Check.objects.filter(service=self.service.id)
+        statuses = [check.status for check in checks]
+        if all(status == 'OK' for status in statuses):
+            self.service.status = 'OK'
+
+        elif 'FAIL' or 'WARNING' in statuses:
+            if 'FAIL' not in statuses:
+                self.service.status = 'WARNING'
+            else:
+                self.service.status = 'FAIL'
```
Comment (hutsulyk) [inline moninag/check/models.py]: Could be simplified
```
if 'FAIL' in statuses:
    self.service.status = 'FAIL'
elif 'WARNING' in statuses:
    self.service.status = 'WARNING'
else:
    self.service.status = 'OK'
```

**Review by hutsulyk — APPROVED** [review]

File: moninag/check/models.py (hunk 10 above)
Comment (ghost) [reply moninag/check/models.py]: done

**Pull request merged**